packages/frontend/src/model/model_editor.tsx

```diff
@@ -76,6 +77,19 @@ export function ModelPane(props: {
     invariant(theories, "Library of theories should be provided as context");
 
     const liveDoc = () => props.liveModel.liveDoc;
+    const formalCells = () =>
+        liveDoc()
+            .doc.notebook.cells.filter((cell) => cell.tag === "formal")
+            .map((cell) => {
+                if (cell.tag === "formal" && cell.content.tag === "object") {
+                    return cell.content.obType.content.toString();
+                }
+                if (cell.tag === "formal" && cell.content.tag === "morphism") {
+                    const j = JSON.parse(JSON.stringify(cell.content.morType.content.valueOf()));
+                    return typeof j === "string" ? j : j.content.toString();
+                }
+                return "";
```

Comment:
Echoing my comment above, this coercion to strings and to/from JSON is a code smell. I'm hoping that we can eliminate all of this.

Reply:
Ok this one is now gone!